### DISCUSSION TITLE
Update node from modal not effective until clicked...

### DISCUSSION THREAD

**Discussion by CSworenI** [xyflow/xyflow#4524] [Q&A] [answered]
I have a custom decision node/edge set that default to yes/no branches (green/red respectively). The starting color of the decision node is split cleanly down the middle for green and red (custom class called 'undecided'). When an edge is chosen as the path forward ('selected' attribute), the 'decision' node should turn the appropriate color.

In my use case, I am persisting `setNodes` to the modal used for updating the edge (and soon the node directly), but the change isn't propagating until I click the node that was changed. What more should I be using in order for the node in the grid to update without having to be clicked on to see the changes?

Here is how I'm declaring the `EdgeModal` (when an edge has been clicked):
```
{clickedEdge ?
	<EdgeModal
		edge={clickedEdge} clearEdge={() => setClickedEdge(null)} nodes={nodes}
		setEdges={setEdges} setNodes={setNodes} ref={edgeModalRef}
	/>
: ''}
```

This is my save function inside the modal:
```
const saveChanges = () => {
	const style = {
		...edge.style, strokeWidth: selected || active ? 6 : 3,
		stroke: isDecision ? COLORS[branch.toLowerCase()] : COLORS[source.data.status], strokeDasharray: active ? '5, 5' : 'none',
		animation: active ? 'dashdraw 0.5s linear infinite' : 'none'
	};
	if (isDecision) {
		updateEdge(edge.id, { ...edge, active, selected, animated: active, label: branch, markerEnd: ARROWS[branch], style });
		source.data.class = selected ? branch.toLowerCase() : 'undecided';
		setNodes((n) => {
			if (n.id === source.id) { return source; }
			else { return n; }
		});
	} else { updateEdge(edge.id, { ...edge, active, animated: active, markerEnd: ARROWS.process, style }); }
	closeModal();
};
```
`active` pertains to whether it is the current execution branch => it should have the `animation` attribute where the edge is in the color of the branch (yes/no) or pertaining to the status of the source node
`selected` pertains to whether the yes/no branch has been chosen as execution path => for the decision edge only; it pertains to the edge having a bolder line to be obvious for users what the decision was at this stage of the flow execution

**Comment by Santoshsohani** [comment]
@CSworenI 
Assume I have a custom node, When I click on a modal component is opened with multiple options, say red and green
If I click on option red the Node should change its color to red and same for green as well.

Now, if the user selects the option red the Node should turn to color red immediately, instead when clicked on the node, The color is changing to red.

Is this the right, Correct me if wrong.

**Reply by CSworenI** [reply]
Yes, that's correct.

My assumption was that using `setNodes` would trigger a rerender on the `ReactFlow` component, but it doesn't. As I tested with the other response, I can completely remove the `setNodes` statement and have the exact same functionality (source node changes after it has been clicked). Do you know if/what the rerender trigger function is and how I might implement it in my modal?

**Comment by bcakmakoglu** [comment]
This part shouldn't do a mutation on the node object.

```ts
// do not mutate node objects when trying to do an update!
source.data.class = selected ? branch.toLowerCase() : 'undecided';
setNodes((n) => {
	if (n.id === source.id) { return source; }
	else { return n; }
});
```

but create a new node object instead:

```ts
setNodes((n) => {
	if (n.id === source.id) {
		return {
			...n,
			data: {
				...n.data,
				class: selected ? branch.toLowerCase() : 'undecided'
			}
		}
	}
	
	return n;
});
```

**Reply by CSworenI** [reply]
I'm seeing the same effect here. It's changing but, again, only after clicking the node in question. It doesn't seem to matter if I mutate or replace the node in the list.

My assumption is that there's some trigger function that forces the grid to update its view of the node, similar to the `updateEdge` function forces it to update its view of the changed edge. I just completely removed the `setNodes` function entirely and witness the same behavior, but if I remove the modification of `source`, then it doesn't change at all.

**Reply by bcakmakoglu** [reply]
The trigger would be that if you click it, a change is applied (a selection change), which would update the node object properly -> your changes suddenly appear.

**Reply by bcakmakoglu** [reply, marked as answer]
Here's an example sandbox: https://codesandbox.io/p/sandbox/serverless-tdd-4xshhs?file=%2Fsrc%2FModal.tsx%3A103%2C52

You can open a modal on the top-left, then use either the update node button or the mutate node button.
Note that the update button will create a new node object at which point the nodes' border color is changed, whereas the mutate button will do what you've done and try to mutate the original node object and pass it back in `setNodes`, which will *not* update the border color unless the node is interacted with (which will cause the object to be updated properly as I said before).

**Reply by CSworenI** [reply]
That seems to work, using `map` on the list of nodes that `setNodes` gives as a parameter. I wish there was a O(1) way of doing it that facilitated the use of nodes in the array, though.

**Reply by bcakmakoglu** [reply]
You could use `updateNodeData` instead: https://codesandbox.io/p/sandbox/hardcore-julien-8lrzlt?file=%2Fsrc%2FModal.tsx%3A19%2C5

**Reply by CSworenI** [reply]
Sandbox not found.

Next to the API reference for `updateNodeData`, I found `updateNode` from `useReactFlow()` and it works exactly as `updateEdge` does. This seems like it's close to the O(1) solution I was looking for.

**Reply by bcakmakoglu** [reply]
Oh the sandbox was private, it's public now.
Yeah, updateNode is also usable although if you're just going to update the nodes' data object, it's better to use `updateNodeData` imo.

Code looks sth like this:
```ts
  const { updateNodeData } = useReactFlow();
  const [selectedNode, setSelectedNode] = useState<string>("a");

  const updateNodeStyles = (nodeId: string) => {
    updateNodeData(selectedNode, { class: 'some-class' });

    setIsVisible(false);
  };
```
